File: pset6/credit.py
from cs50 import get_string
from sys import exit


def main():

    # Get input from user
    number = get_string("Number: ")
    digits = len(number)

    # Check length is valid
    if digits < 13 or digits > 16 or digits == 14:
        print("INVALID")
        exit(1)

    # Luhn's algorithm
    if luhns(number) is False:
        print("INVALID")
        exit(1)

    # VISA or MASTERCARD
    if digits == 13 or digits == 16:
        # Use the number at the beginning to determine the type
        if number[0] == "4":
            print("VISA")
            exit(0)

        elif (int(number[:2]) > 50) and (int(number[:2]) < 56):
            print("MASTERCARD")
            exit(0)

        else:
            print("INVALID")
            exit(0)

    # AMEX
    elif digits == 15:
        # Check the beginning of the number is correct
        if number[:2] == "34" or number[:2] == "37":
            print("AMEX")
            exit(0)
        else:
            print("INVALID")
            exit(1)


# Function takes the number (as a string) as an input, and returns a boolean value
# which indicates whether the number is or is not a valid credit card number
def luhns(number):

    even_digits = []
    odd_digits = []

    for i in range(len(number)):
        # If even, append digit in int form to odds list
        if i % 2 == 0:
            odd_digits.append(int(number[len(number) - i - 1]))
        # If not, append to evens list
        else:
            even_digits.append(int(number[len(number) - i - 1]))

    # Multiply all entries in even_digits by 2
    even_digits = [i * 2 for i in even_digits]

    # If we have a two digit number, preemptively sum the digits
    for k in range(len(even_digits)):
        # Can be at most two digits
        if even_digits[k] > 9:
            even_digits[k] = 1 + (even_digits[k] - 10)

    # Now we can just sum over the two lists
    output = sum(even_digits) + sum(odd_digits)

    if output % 10 == 0:
        return True

    else:
        return False


main()

File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int find_length(long num);
int extract_digit(long num, int digit_position);
bool luhns_alg(long n);
void check_type(long n);
void test(long n);



int main(void)
{
    long number = get_long("Number: ");
    test(number);
}



void test(long n)
{
    //First check the length - must be between 13 and 16
    if (n > pow(10, 16))
    {
        printf("INVALID\n");
    }
    else if (n < pow(10, 12))
    {
        printf("INVALID\n");
    }
    else
    {
        //Now check validity
        bool validity = luhns_alg(n);

        if (validity == false)
        {
            printf("INVALID\n");
        }
        else
        {
            //Finally check type
            check_type(n);
        }
    }
}


//Check the beginning digits to determine type and double check
//length is correct
void check_type(long n)
{
    int len = find_length(n);
    int first_digit = extract_digit(n, len - 1);
    int second_digit = extract_digit(n, len - 2);

    //Amex start with 34 or 37 and are 15 digits long
    if (first_digit == 3 && len == 15)
    {
        if (second_digit == 4 || second_digit == 7)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //Mastercard start with 51, 52, 53, 54, 55 and are 16 digits long
    else if (first_digit == 5 && len == 16)
    {
        if (second_digit < 6)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    //Visa start with 4 andare 13 or 16 digits long
    else if (first_digit == 4)
    {
        if (len == 13 || len == 16)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}


//Implement Luhn's Algorithm
bool luhns_alg(long n)
{
    int len = find_length(n);

    //Multiply even digits by two and sum the digits of these
    int even_sum = 0;

    for (int i = 1; i < len; i = i + 2)
    {
        int dig = extract_digit(n, i) * 2;

        //Check if single or double digits (must be <20)
        if (dig < 10)
        {
            even_sum = even_sum + dig;
        }
        else
        {
            int dig_sum = (dig % 10) + ((dig / 10) % 10);
            even_sum = even_sum + dig_sum;
        }
    }

    //Sum the odd digits
    int odd_sum = 0;

    for (int j = 0; j < len; j = j + 2)
    {
        int dig = extract_digit(n, j);
        odd_sum = odd_sum + dig;
    }

    //Add these two together
    int sum = even_sum + odd_sum;

    //Modulo 10
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}


//Extract specific digits
int extract_digit(long num, int digit_position)
{
    long divisor = pow(10, digit_position);
    int digit = (num / divisor) % 10;
    return digit;
}


//Find length between 13 and 16 digits; input is a number with between 13 and 16 digits
int find_length(long num)
{
    if (num < pow(10, 13))
    {
        return 13;
    }
    else if (num < pow(10, 14))
    {
        return 14;
    }
    else if (num < pow(10, 15))
    {
        return 15;
    }
    else
    {
        return 16;
    }
}